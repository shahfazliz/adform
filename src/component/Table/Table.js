import React from 'react';
import BootstrapTable from 'react-bootstrap/Table';
import PropTypes from 'prop-types';

import '../../css/bootstrap.min.css';

const setHeader = header => header.map((head, index) => <th key={ index }>{ head.name }</th>);

const setValues = (row, header) => header.map((head, index) => <td key={ index }>{ row[head.key] }</td>);

const setRows = (items, header) => items
  .length
  ? items.map((row, index) => <tr key={ index }>{ setValues(row, header) }</tr>)
  : <tr>
    <td colSpan={ header.length }>Sorry, no data available</td>
  </tr>;

/**
 * Using adapter design pattern, we wrap the bootstrap Table component into our own custom table
 * component. This design pattern allows us to change the underlying third party component without
 * changing our code base.
 *
 * @param {Object[]} props.header - An array of object that will be the header of the table. This
 *                                  parameter is required because the sequence of keys in an object
 *                                  is not guaranteed.
 *
 * @param {String} props.header.key - The key of the associated object in props.item
 *
 * @param {String} props.header.name - The name of the header to display
 *
 * @param {Object[]} props.items -  An array of objects that should be the content of the table where
 *                                  the keys are the header of the table and the values are the cell
 *                                  values.
 * @example
 * <Table
 *  header={[{
 *    key: 'header1',
 *    name: 'Header One',
 *  }, {
 *    key: 'header2',
 *    name: 'Header Two',
 *  }, {
 *    key: 'header3',
 *    name: 'Header Three',
 *  }]}
 *
 *  items={[{
 *    header1: 'value1a',
 *    header2: 'value2a',
 *    header3: 'value3a',
 *  }, {
 *    header1: 'value1b',
 *    header2: 'value2b',
 *    header3: 'value3b',
 *  }, {
 *    header1: 'value1c',
 *    header2: 'value2c',
 *    header3: 'value3c',
 * }]} />
 *
 * @returns {React.Component}
 */
function Table(props) {
  return (<BootstrapTable striped bordered hover>
    <thead>
      <tr>{ setHeader(props.header) }</tr>
    </thead>
    <tbody>{ setRows(props.items, props.header) }</tbody>
  </BootstrapTable>);
}

Table.propTypes = {
  header: PropTypes
    .arrayOf(PropTypes.object)
    .isRequired,
  items: PropTypes
    .arrayOf(PropTypes.object)
    .isRequired,
};

export default Table;