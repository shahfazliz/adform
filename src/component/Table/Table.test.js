import React from 'react';
import { configure, shallow } from 'enzyme';

import Adapter from 'enzyme-adapter-react-16';
import Table from './Table';

configure({ adapter: new Adapter() });

it('should render object header and object items in a table, if there is data', () => {
  const tableComponent = shallow(<Table
    header={[{
      key: 'header1',
      name: 'Header One',
    }, {
      key: 'header2',
      name: 'Header Two',
    }, {
      key: 'header3',
      name: 'Header Three',
    }]}

    items={[{
      header1: 'value1a',
      header2: 'value2a',
      header3: 'value3a',
    }, {
      header1: 'value1b',
      header2: 'value2b',
      header3: 'value3b',
    }, {
      header1: 'value1c',
      header2: 'value2c',
      header3: 'value3c',
    }]} />);

  expect(tableComponent.find('thead')).toMatchSnapshot(
    `<thead>
      <tr>
        <th>Header One</th>
        <th>Header Two</th>
        <th>Header Three</th>
      </tr>
    </thead>`);

  expect(tableComponent.find('tbody')).toMatchSnapshot(
    `<tbody>
      <tr>
        <td>value1a</td>
        <td>value2a</td>
        <td>value3a</td>
      </tr>
      <tr>
        <td>value1b</td>
        <td>value2b</td>
        <td>value3b</td>
      </tr>
      <tr>
        <td>value1c</td>
        <td>value2c</td>
        <td>value3c</td>
      </tr>
    </tbody>`);
});

it('should render object header but no items in a table, when there is no data', () => {
  const tableComponent = shallow(<Table
    header={[{
      key: 'header1',
      name: 'Header One',
    }, {
      key: 'header2',
      name: 'Header Two',
    }, {
      key: 'header3',
      name: 'Header Three',
    }]}

    items={[]} />);

  expect(tableComponent.find('thead')).toMatchSnapshot(
    `<thead>
      <tr>
        <th>Header One</th>
        <th>Header Two</th>
        <th>Header Three</th>
      </tr>
    </thead>`);

  expect(tableComponent.find('tbody')).toMatchSnapshot(
    `<tbody>
      <tr>
        <td col-span>Sorry, no data available</td>
      </tr>
    </tbody>`);
});