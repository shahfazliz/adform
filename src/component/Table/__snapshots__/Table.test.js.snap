// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render object header and object items in a table, if there is data: <tbody>
      <tr>
        <td>value1a</td>
        <td>value2a</td>
        <td>value3a</td>
      </tr>
      <tr>
        <td>value1b</td>
        <td>value2b</td>
        <td>value3b</td>
      </tr>
      <tr>
        <td>value1c</td>
        <td>value2c</td>
        <td>value3c</td>
      </tr>
    </tbody> 1`] = `ShallowWrapper {}`;

exports[`should render object header and object items in a table, if there is data: <thead>
      <tr>
        <th>Header One</th>
        <th>Header Two</th>
        <th>Header Three</th>
      </tr>
    </thead> 1`] = `ShallowWrapper {}`;

exports[`should render object header and object items in a table: <tbody>
      <tr>
        <td>value1a</td>
        <td>value2a</td>
        <td>value3a</td>
      </tr>
      <tr>
        <td>value1b</td>
        <td>value2b</td>
        <td>value3b</td>
      </tr>
      <tr>
        <td>value1c</td>
        <td>value2c</td>
        <td>value3c</td>
      </tr>
    </tbody> 1`] = `ShallowWrapper {}`;

exports[`should render object header and object items in a table: <thead>
      <tr>
        <th>Header One</th>
        <th>Header Two</th>
        <th>Header Three</th>
      </tr>
    </thead> 1`] = `ShallowWrapper {}`;

exports[`should render object header but no items in a table, when there is no data: <tbody>
      <tr>
        <td col-span>Sorry, no data available</td>
      </tr>
    </tbody> 1`] = `ShallowWrapper {}`;

exports[`should render object header but no items in a table, when there is no data: <tbody>
      <tr>
        <td colSpan={3}>Sorry, no data available</td>
      </tr>
    </tbody> 1`] = `ShallowWrapper {}`;

exports[`should render object header but no items in a table, when there is no data: <tbody>
      <tr>
        <td colSpan={4}>Sorry, no data available</td>
      </tr>
    </tbody> 1`] = `ShallowWrapper {}`;

exports[`should render object header but no items in a table, when there is no data: <tbody>
      <tr>
        <td>Sorry, no data available</td>
      </tr>
    </tbody> 1`] = `ShallowWrapper {}`;

exports[`should render object header but no items in a table, when there is no data: <thead>
      <tr>
        <th>Header One</th>
        <th>Header Two</th>
        <th>Header Three</th>
      </tr>
    </thead> 1`] = `ShallowWrapper {}`;
